packages/core/src/browser/drag-and-drop.ts

```diff
@@ -0,0 +1,137 @@
+/*
+* Copyright (C) 2017 Ericsson and others.
+*
+* Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+*/
+import { injectable, inject } from "inversify";
+import { TreeWidget, TreeProps, ITreeModel, ITreeNode, NodeProps } from './tree';
+import { ContextMenuRenderer } from "./context-menu-renderer";
+import { ElementAttrs } from "@phosphor/virtualdom";
+
+export const DROPZONE_CLASS = 'dropzone';
+
+@injectable()
+export class DragAndDraggable extends TreeWidget {
+
+    constructor(
+        @inject(TreeProps) readonly props: TreeProps,
+        @inject(ITreeModel) readonly model: ITreeModel,
+        @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer
+    ) {
+        super(props, model, contextMenuRenderer);
+    }
+
+    protected createNodeAttributes(node: ITreeNode, props: NodeProps): ElementAttrs {
+        const elementAttrs: ElementAttrs = super.createNodeAttributes(node, props);
+        const childExpansion = this.isExandable(node) ? 0 : 1;
+
+        if (!childExpansion) {
+            // Folder, so it could be a dropzone
+            return {
+                ...elementAttrs,
+                ondragenter: event => this.handleDragEnterEvent(node, event),
+                ondrop: event => this.handleDropEvent(node, event),
+                ondragover: event => this.handleDragOverEvent(node, event),
+                ondragend: event => console.log("END"),
+                ondragexit: event => console.log("EXIT"),
+                ondragleave: event => {
+                    const target = <HTMLElement>(event.target as {});
+                    const parent = <HTMLDivElement>(target.parentElement as {});
+                    parent.classList.remove("activeDropZone");
+                },
+            };
+        }
+        return elementAttrs;
+    }
+
+    protected handleDragEnterEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = <HTMLElement>(event.target as {});
+        const parent = <HTMLDivElement>(target.parentElement as {});
+        parent.classList.add("activeDropZone");
+    }
+    protected handleDragOverEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+    }
+
+    protected handleDropEvent(node: ITreeNode, event: DragEvent): void {
+        event.stopPropagation();
+        event.preventDefault();
+        event.dataTransfer.dropEffect = 'copy'; // Explicitly show this is a copy.
+        const target = <HTMLElement>(event.target as {});
+        const parent = <HTMLDivElement>(target.parentElement as {});
+        parent.classList.remove("activeDropZone");
+        this.drop_handler(event, node);
+    }
+
+    protected drop_handler(ev: DragEvent, node: ITreeNode) {
```

Comment:
please follow naming conventions for functions and methods: https://github.com/theia-ide/theia/wiki/Coding-Guidelines#names
> Use camelCase for function and method names

Reply:
It looks like the whole function can be replaced by a service, something like `UploadService.upload(data, destination)`.